plugins/push/frontend/public/javascripts/countly.models.js

```diff
@@ -192,12 +200,80 @@
                     sent: pushNotificationDtoItem.result.sent,
                     actioned: pushNotificationDtoItem.result.actioned,
                     createdBy: pushNotificationDtoItem.creator,
-                    platform: self.mapPlatforms(pushNotificationDtoItem.platforms),
+                    platforms: self.mapPlatforms(pushNotificationDtoItem.platforms),
                     content: pushNotificationDtoItem.messagePerLocale.default,
                 };
             });
             return rowsModel;
         },
+        mapMessageButtons: function(numberOfButtons, dto) {
+            if (numberOfButtons === 1) {
+                return [{label: dto.messagePerLocale["default|0|t"], url: dto.messagePerLocale["default|0|l"]}];
+            }
+            if (numberOfButtons === 2) {
+                return [
+                    {label: dto.messagePerLocale["default|0|t"], url: dto.messagePerLocale["default|0|l"]},
+                    {label: dto.messagePerLocale["default|1|t"], url: dto.messagePerLocale["default|1|l"]},
+                ];
+            }
+            return [];
+        },
+        mapEndDate: function(type, dto) {
+            if (type === this.TypeEnum.AUTOMATIC) {
+                return moment(dto.autoEnd).format("MMMM Do, YYYY, H:mm");
+            }
+            return "never";
+        },
+        mapDeliveryType: function() {
+            //TODO: map push notification delivery type
+            return null;
+        },
+        mapAudienceSelectionType: function() {
+            //TODO: map push notification audience selection type
+            return null;
+        },
+        mapErrors: function(dto) {
+            //TODO: map push notification message errors;
+            return {codes: dto.result.errorCodes, messages: dto.result.error};
+        },
+        mapPushNotificationDtoToModel: function(dto) {
+            var self = this;
```

Comment:
These "Model"s are Data structures which are defined in push, right? I mean you use a format in frontend (which is carried by a DTO btw) and you convert them to backend's when necessary, no?

Reply:
The model object is designed and used throughout the push plugin in the client side (after being mapped to from dto, the incoming transfer object received from the server). 
If we need outgoing mappers (converting the model to a specific dto) when making a server request than we will have to write them as well.